vllm/engine/llm_engine.py

```diff
@@ -558,14 +564,25 @@ def step(self) -> List[RequestOutput]:
         if scheduler_outputs.is_empty():
             return ignored
 
+        draft_tokens = None
+        if self.spec_handler:
+            draft_tokens = self.spec_handler.propose()
+        
         # Execute the model.
         output = self._run_workers(
             "execute_model",
             seq_group_metadata_list=seq_group_metadata_list,
             blocks_to_swap_in=scheduler_outputs.blocks_to_swap_in,
             blocks_to_swap_out=scheduler_outputs.blocks_to_swap_out,
             blocks_to_copy=scheduler_outputs.blocks_to_copy,
+            draft_tokens=draft_tokens
         )
+        
+        if self.spec_handler:
+            self.spec_handler.accept(output)
```

Comment:
I think spec_handler should be stateless (besides potentially managing its own KV cache), as I think that it is a good idea to centralize state as much as possible, so perhaps have it return n (following the original algorithm from https://arxiv.org/abs/2211.17192 ), and then SequenceMetadata is updated to invalidate the cache